### PR TITLE
[batch] attributes table

### PR DESCRIPTION
**Pull request by danking**
FOR TESTING ONLY

### CODE REVIEW

File: batch/test-locally.sh [hunk 1]
```diff
@@ -39,7 +39,8 @@ fi
 export JOBS_TABLE=jobs-$(../generate-uid.sh)
 export JOBS_PARENTS_TABLE=jobs-parents-$(../generate-uid.sh)
 export BATCH_TABLE=batch-$(../generate-uid.sh)
-tables=($JOBS_TABLE $JOBS_PARENTS_TABLE $BATCH_TABLE)
+export BATCH_ATTRIBUTES_TABLE=batch-attributes-$(../generate-uid.sh)
```
Comment (jigold) [inline batch/test-locally.sh]: I don't think this script is used anymore. We should clean up the test-locally infrastructure in another PR.

File: batch/test/test_database.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import os
+import unittest
+
+from batch.database import Database, run_synchronous
+
+
+class Test(unittest.TestCase):
+    def setUp(self):
+        self.db = Database.create_synchronous(os.environ.get('CLOUD_SQL_CONFIG_PATH'))
```
Comment (jigold) [inline batch/test/test_database.py]: I deleted this file because we don't have a way to make new tables with the permissions on the current batch setup

File: batch/create-batch-tables.sql [hunk 3]
```diff
@@ -86,3 +86,12 @@ CREATE TRIGGER trigger_jobs_update AFTER UPDATE ON jobs
 $$
 
 DELIMITER ;
+
+CREATE TABLE IF NOT EXISTS `batch-attributes` (
+  `batch_id` BIGINT NOT NULL,
+  `key` TEXT(65535) NOT NULL,
+  `value` TEXT(65535),
+  PRIMARY KEY (`batch_id`)
```
Comment (jigold) [inline batch/create-batch-tables.sql]: Do you want the key to be `batch_id`, `key`? Maybe not if you want to grab all the attributes by batch.

File: batch/batch/server/server.py [hunk 4]
```diff
@@ -847,29 +847,26 @@ def is_successful(self):
...
-
-    return jsonify([await batch.to_dict(include_jobs=False) for batch in batches])
+    complete = params.get('complete')
+    if complete:
+        del params['complete']
+        complete = complete == '1'
+    success = params.get('success')
+    if success:
+        del params['success']
+        success = success = '1'
```
Comment (jigold) [inline batch/batch/server/server.py]: `success = success == '1'`

File: batch/batch/server/database.py [hunk 5]
```diff
@@ -151,3 +195,16 @@ def __init__(self, db):
...
+
+class BatchAttributesTable(Table):
+    def __init__(self, db):
+        super().__init__(db, 'batch-attributes')
+
+    async def _query(self, *select, **where):
+        return await super().get_record(where, select_fields=select)
+
+    async def get_attributes(self, batch_id):
+        return self._query('key', 'value', batch_id=batch_id)
```
Comment (jigold) [inline batch/batch/server/database.py]: I think you need awaits here and below.

File: batch/batch/server/database.py [hunk 6]
```diff
@@ -135,6 +136,49 @@ def __init__(self, db):
...
+
+        values += user
+        wheres += "job.user = %s"
+        if deleted is not None:
+            if deleted:
+                wheres += "batch.deleted"
+            else:
+                wheres += "not batch.deleted"
+        if complete is not None:
+            condition = "not batch.is_open and batch.n_completed = batch.n_jobs"
```
Comment (jigold) [inline batch/batch/server/database.py]: The  `is_open` will conflict with my atomic batch creation PR. I bet yours will go in first though so I'll end up changing it later.

File: batch/batch/server/database.py [hunk 7]
```diff
@@ -135,6 +136,49 @@ def __init__(self, db):
...
+                wheres += condition
+            else:
+                wheres += f"not ({condition})"
+        if attributes:
+            joins += "inner join {self._db.batch_attributes.name} as attr using (batch_id)"
+            groups += "batch.id"
+            for k, v in attributes.items():
+                values += v
+                values += len(attributes)
+                havings += f"sum(attr.`{k}` = %s) = %s"
```
Comment (jigold) [inline batch/batch/server/database.py]: I don't see how this line is correct. Shouldn't the second `%s` be 1 and not `len(attributes)`?

File: batch/Makefile [hunk 8]
```diff
@@ -13,7 +13,7 @@ BATCH_TEST_IMAGE = gcr.io/$(PROJECT)/batch-test:$(shell docker images -q --no-tr
 
 PY_FILES = $(shell find batch -iname \*.py -not -exec git check-ignore -q {} \; -print)
 
-PYTHONPATH := $${PYTHONPATH:+$${PYTHONPATH}:}../hailjwt
+PYTHONPATH := ../hailjwt$${PYTHONPATH:+:$${PYTHONPATH}}
```
Comment (jigold) [inline batch/Makefile]: Why is this necessary?

File: batch/test/test_database.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            assert len(records) == 0
+        finally:
+            self.db.drop_table_sync(t.name)
+
+    def test_has_record(self):
+        t = self.temp_table()
+        try:
+            assert run_synchronous(t.has_record({'id': 4}))
+        finally:
+            self.db.drop_table_sync(t.name)
```
Comment (jigold) [inline batch/test/test_database.py]: delete file!

File: batch/Makefile (hunk 8 above)
Comment (danking) [reply batch/Makefile]: just seems right. I don't want to pick up a hailjwt that was accidentally pip installed. This is all irrelevant anyway with build.yaml

File: batch/create-batch-tables.sql (hunk 3 above)
Comment (danking) [reply batch/create-batch-tables.sql]: I want to map from batch_id to its attributes (ergo the key) and from a key-value pair to a batch_id (ergo the index).

**Comment by jigold** [comment]
I don't think you ever insert records into the db in this PR.
